aip/0153.md

```diff
@@ -0,0 +1,164 @@
+  option (google.api.http) = {
+    post: "/v1/{parent=publishers/*}/books:import"
+    body: "*"
+  };
+  option (google.longrunning.operation_info) = {
+    response_type: "ImportBooksResponse"
+    metadata_type: "ImportBooksMetadata"
+  };
+}
+```
+
+- The method **must** return a long-running operation (see [AIP-151][]) unless
+  the service can guarantee that it will _never_ need more than a few seconds
+  to complete.
+- The HTTP verb **must** be `POST`, and the `body` **must** be `"*"`.
+- A `parent` field **should** be included as part of the URI.
+  - If importing into or exporting from multiple resources is required, the API
+    **should** keep the `parent` field and allow the user to use the `-`
+    character to indicate multiple parents (see [AIP-159][]).
+  - On import, if the user provides a specific parent, the API **must** reject
+    any imported resources that would be added to a different parent.
+- The URI suffix **should** be `:import` or `:export`.
+
+### Importing data into a resource
+
+Services **should** implement a common pattern for importing data into a single
+resource:
+
+```proto
+rpc ImportPages(ImportPagesRequest) returns (google.longrunning.Operation) {
+  option (google.api.http) = {
+    post: "/v1/{book=publishers/*/books/*}:importPages"
+    body: "*"
+  };
+  option (google.longrunning.operation_info) = {
+    response_type: "ImportPagesResponse"
+    metadata_type: "ImportPagesMetadata"
+  };
+}
+```
+
+- The method **must** return a long-running operation (see [AIP-151][]) unless
+  the service can guarantee that it will _never_ need more than a few seconds
+  to complete.
+- The HTTP verb **must** be `POST`, and the `body` **must** be `"*"`.
+- A field representing the resource that data is being imported into **should**
+  be included as part of the URI. The field **should** be named after the
+  resource (and **should not** be called `name`).
+- The URI suffix should include both the verb and a noun for the data itself,
+  such as `:importPages` or `:exportPages`.
+
+### Request object
+
+Imports and exports often require two fundamentally different types of
+configuration:
+
+1. Configuration specific to the source or destination.
+2. Congiguration regarding the imported or exported data itself.
+
+Source or destination configuration should be grouped into a single message and
+placed inside a oneof:
+
+```proto
+message ImportBooksRequest {
+  string parent = 1;
+  oneof source {
+    AuthorSource author_source = 2;
+    TranslatorSource translator_source = 3;
+  }
+  string isbn_prefix = 4;
+}
+```
+
+- The source configuration messages **must** be placed within a `oneof source`
+  (for import) or `oneof destination` (for export), even if there is only one.
+  (This maintains flexibility to add more later.)
+- Configuration related to the data itself (and therefore common across all
+  sources) **must** be placed at the top-level of the request message.
+
+### Common sources
```

Comment:
I'm somewhat uncomfortable with Google-specific advice in a non-Google-specific AIP. But it's better to have it to start with and then move it than not have it at all, I guess.

Reply:
I accept the point as valid. Perhaps a Google Cloud AIP for "common import/export sources", which this could link to?

Reply:
Yes, I think that would be the best solution.